pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,483 @@ func (c *Controller) processNextWorkItem() bool {
-	tenant.Status.Message = utils.Trim(err.Error())
-	_, updateStatusErr := c.updateStatus(tenant)
-	return updateStatusErr
+func (c *Controller) checkNamespaceExists(name string) (bool, error) {
+	namespaces := c.factory.CoreV1().Namespaces()
+	namespace, err := namespaces.Get(name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return false, nil
+		}
+		err = errors.WithMessagef(err, "error: failed to get namespace %q", name)
+		return false, err
+	}
+	return namespace.GetDeletionTimestamp().IsZero(), nil
 }
 
-func (c *Controller) rollbackIfRequired(tenantKey string) {
-	tenant, err := c.fetcher.ByKey(tenantKey)
+func (c *Controller) createTenantNamespace(config clientConfig, tenant *api.Tenant) (string, error) {
+	c.logPrintln(tenant, "creating new tenant namespace")
+	namespaceManager := c.getNamespaceManager(config)
+	nsName, err := namespaceManager.Create(tenant.GetName(), nil)
 	if err != nil {
-		log.Printf("ERROR: Could not get tenant during rollback: %s", err.Error())
+		err = errors.WithMessage(err, "failed to create new tenant namespace")
+		c.logPrintln(tenant, err)
+		return "", err
 	}
-	if tenant.Status.Progress != api.TenantProgressFinished {
-		_ = c.rollback(tenant)
+	return nsName, err
+}
+
+func (c *Controller) deleteTenantNamespace(namespace string, tenant *api.Tenant, config clientConfig) error {
+	if namespace == "" {
+		return nil
 	}
+	c.logPrintf(tenant, "rolling back tenant namespace %q", namespace)
+	namespaceManager := c.getNamespaceManager(config)
+	err := namespaceManager.Delete(namespace)
+	if err != nil {
+		err = errors.WithMessagef(err, "failed to delete tenant namespace %q", namespace)
+		c.logPrintln(tenant, err)
+		return err
+	}
+	return nil
 }
 
-func (c *Controller) rollback(tenant *api.Tenant) error {
-	log.Printf("Rollback tenant %s", tenant.GetName())
-	if tenant.Status.TenantNamespaceName == "" {
-		log.Printf("Nothing to rollback for tenant %s", tenant.GetName())
-	} else {
-		err := c.deleteNamespace(tenant)
+/*
+reconcileTenantRoleBinding compares the actual state of the role binding
+in the tenant namespace with the desired state.
+In case of a mismatch it tries to achieve the desired state by replacing
+the existing role binding resource object(s) by a new one.
+
+Output parameter `updateNeeded` indicates whether the need for an update
+has been detected, and will be returned both in case of success and error.
+In case of success it indicates whether an update has been performed.
+In case of error a value of `true` indicates that an update would have to
+be performed (and maybe has been done partially), while a value of 'false'
+does NOT mean that no update is necessary BUT that in the moment the error
+occurred the need for an update has not been detected (a subtle but yet
+important difference).
+*/
+func (c *Controller) reconcileTenantRoleBinding(tenant *api.Tenant, namespace string, config clientConfig) (updateNeeded bool, err error) {
```

Comment:
Result should be named updated or changed as the change was already done when the function returns.

Reply:
Actually not.

I've tried to clarify the meaning if the output parameter in the method doc:

> Output parameter `updateNeeded` indicates whether the need for an update
> has been detected, and will be returned both in case of success and error.
> In case of success it indicates whether an update has been performed.
> In case of error a value of `true` indicates that an update would have to
> be performed (and maybe has been done partially), while a value of 'false'
> does NOT mean that no update is necessary BUT that in the moment the error
> occurred the need for an update has not been detected (a subtle but yet
> important difference).

Did I miss anything?

Reply:
Improved in 85b310470.

Renamed `updateNeeded` to `needForUpdateDetected` to express more clearly that this is an informative flag for the caller, not a requirement.